rust/src/array.rs

```diff
@@ -15,297 +15,421 @@
+/// Array of List<T>
+pub struct ListArray<T: ArrowPrimitiveType> {
+    len: usize,
+    data: List<T>,
+    null_count: usize,
     validity_bitmap: Option<Bitmap>,
-    /// The array of elements
-    data: ArrayData,
 }
 
-impl Array {
-    /// Create a new array where there are no null values
-    pub fn new(len: usize, data: ArrayData) -> Self {
-        Array {
-            len: len as i32,
-            data,
+impl<T> ListArray<T>
+where
+    T: ArrowPrimitiveType,
+{
+    pub fn get(&self, i: usize) -> &[T] {
+        self.data.get(i)
+    }
+
+    pub fn list(&self) -> &List<T> {
+        &self.data
+    }
+}
+
+/// Create a ListArray<T> from a List<T> without null values
+impl<T> From<List<T>> for ListArray<T>
+where
+    T: ArrowPrimitiveType,
+{
+    fn from(list: List<T>) -> Self {
+        let len = list.len();
+        ListArray {
+            len,
+            data: list,
             validity_bitmap: None,
             null_count: 0,
         }
     }
+}
 
-    /// Get a reference to the array data
-    pub fn data(&self) -> &ArrayData {
-        &self.data
+/// Create ListArray<u8> from Vec<&'static str>
+impl From<Vec<&'static str>> for ListArray<u8> {
+    fn from(v: Vec<&'static str>) -> Self {
+        let mut builder: ListBuilder<u8> = ListBuilder::with_capacity(v.len());
+        for s in v {
+            builder.push(s.as_bytes())
+        }
+        ListArray::from(builder.finish())
     }
+}
 
-    /// number of elements in the array
-    pub fn len(&self) -> usize {
-        self.len as usize
+/// Create ListArray<u8> from Vec<String>
+impl From<Vec<String>> for ListArray<u8> {
+    fn from(v: Vec<String>) -> Self {
+        let mut builder: ListBuilder<u8> = ListBuilder::with_capacity(v.len());
+        for s in v {
+            builder.push(s.as_bytes())
+        }
+        ListArray::from(builder.finish())
     }
+}
 
-    /// number of null elements in the array
-    pub fn null_count(&self) -> usize {
-        self.null_count as usize
+impl<T> Array for ListArray<T>
+where
+    T: ArrowPrimitiveType,
+{
+    fn len(&self) -> usize {
+        self.len
     }
-
-    /// If null_count is greater than zero then the validity_bitmap will be Some(Bitmap)
-    pub fn validity_bitmap(&self) -> &Option<Bitmap> {
+    fn null_count(&self) -> usize {
+        self.null_count
+    }
+    fn validity_bitmap(&self) -> &Option<Bitmap> {
         &self.validity_bitmap
     }
+    fn as_any(&self) -> &Any {
+        self
+    }
 }
 
-macro_rules! array_from_primitive {
-    ($DT:ty) => {
-        impl From<Vec<$DT>> for Array {
-            fn from(v: Vec<$DT>) -> Self {
-                Array {
-                    len: v.len() as i32,
-                    null_count: 0,
-                    validity_bitmap: None,
-                    data: ArrayData::from(v),
-                }
-            }
+/// Array of T
+pub struct BufferArray<T: ArrowPrimitiveType> {
```

Comment:
wonder if it's better to call this `PrimitiveArray`?

Reply:
+1 `BufferArray` is a bit misleading